clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1044,6 +1044,46 @@ public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callbac
         return doSend(interceptedRecord, callback);
     }
 
+    /**
+     * Preload metadata for the specified topic to avoid blocking on the first send() call.
+     * This method will block until metadata for the topic is available or until
+     * max.block.ms timeout is reached.
+     * 
+     * <p>For multiple topics, users can call this method concurrently in separate threads
+     * to control parallelism according to their needs.
+     *
+     * @param topic the topic to warm up metadata for
+     * @throws InterruptedException if interrupted while waiting for metadata
+     * @throws TimeoutException if metadata could not be fetched within max.block.ms
+     * @throws KafkaException for other errors
+     */
+    @Override
+    public void warmup(String topic) throws InterruptedException {
+        warmup(topic, maxBlockTimeMs);
+    }
+
+    /**
+     * Preload metadata for the specified topic to avoid blocking on the first send() call.
+     * This method will block until metadata for the topic is available or until
+     * the specified timeout is reached.
+     * 
+     * <p>For multiple topics, users can call this method concurrently in separate threads
+     * to control parallelism according to their needs.
+     *
+     * @param topic the topic to warm up metadata for
+     * @param timeoutMs the maximum time to wait for metadata in milliseconds
+     * @throws InterruptedException if interrupted while waiting for metadata
+     * @throws TimeoutException if metadata could not be fetched within the timeout
+     * @throws KafkaException for other errors
+     */
+    @Override
+    public void warmup(String topic, long timeoutMs) throws InterruptedException {
```

Comment:
Making this a public change seems like a bit of overkill to me. Perhaps you could call `partitionsFor` in your env?

Reply:
open https://github.com/apache/kafka/pull/20842 for review 
close this one.